src/Payments.sol

```diff
@@ -223,6 +223,71 @@ contract Payments is
         return railId;
     }
 
+    function terminateRail(
+        uint256 railId
+    ) external validateRailActive(railId) noRailModificationInProgress(railId) {
+        Rail storage rail = rails[railId];
+
+        // Ensure only the payer can terminate the rail
+        require(
+            rail.from == msg.sender,
+            "only the rail payer can terminate the rail"
+        );
+
+        Account storage payer = accounts[rail.token][rail.from];
+
+        // Settle account lockup
+        settleAccountLockup(payer);
+
+        // Check that the rail is not in debt
+        // (i.e., client has enough funds to pay for services already taken on this rail upto and including the current epoch)
+        require(
+            block.number < payer.lockupLastSettledAt + rail.lockupPeriod,
+            "rail is in debt; cannot terminate"
+        );
+
+        // Calculate the effective remaining lockup period that hasn't been settled
+        require(
+            payer.lockupLastSettledAt <= block.number,
+            "lockup settlement epoch cannot be in the future"
+        );
+        uint256 effectiveLockupPeriod = rail.lockupPeriod -
+            (block.number - payer.lockupLastSettledAt);
+
+        // Calculate funds that can be released (funds locked for future epochs + fixed lockup)
+        // note: this does not include the current epoch as the rail still needs to be paid for for the
+        // current epoch
+        uint256 fundsToRelease = (rail.paymentRate * effectiveLockupPeriod) +
+            rail.lockupFixed;
```

Comment:
I want to avoid locking new funds but, IMO, terminating rails this way shouldn't release locked funds. Terminate is really "SIGKILL". The "nice" way to stop a rail is to terminate all deals on that rail through the rails operator. But, if the rail is buggy, this lets the client say "ok, no more".

Reply:
@Stebalien I see. So on rail termination, we ideally want to keep the future(rail period) funds locked but allows a client to exit the rail _after_ that.

Reply:
@Stebalien We need to discuss this a bit. If we set the rate to zero here, rail settlement wont actually settle anything.

Reply:
Yes. It cuts off the flow of funds to the rail but still preserves all the other guarantees. Otherwise, locking funds is kind of pointless.

An alternative (potential addition?) to terminate is "freeze": the ability to freeze all funds into a rail while still allowing funds to flow out. Eventually the rail would go into debt. I bring this up because it _might_ be useful in some cases where there are disputes that need to be handled out-of-band. To "unfreeze" a rail, the client would have to fully fund it. But, IMO, that's a separate feature we should consider later.

Reply:
@Stebalien I have refactored this to ensure both account lockup and non-arbitrated rails are fully settled before this can happen. For arbitrated rails -> we push the current rate to the queue.

We then set rail rate to 0.

Please take a look.